### PR TITLE
Implemented multiple K's - #23

### PR DESCRIPTION
**Pull request by jan-gerling**
* Extended definition of isStable of PMTracker
 * Adapted and extended PMDatabaseTest class
 * changed commitThreshold to List
 * Additional real project tests
 * Various bug fixes
 * Verified existing Tests

### CODE REVIEW

**Comment by jan-gerling** [comment]
image
image

File: data-collection/src/main/java/refactoringml/ProcessMetricTracker.java [hunk 1]
```diff
@@ -74,6 +77,29 @@ public void resetCounter(CommitMetaData commitMetaData) {
 
 	public ProcessMetrics getCurrentProcessMetrics() { return currentProcessMetrics; }
 
+	public int getCommitCountThreshold() { return currentCommitThreshold; }
+
+	//Filter class files that were not refactored in the last K commits and not already found with a lower K.
+	//TODO: If a class has an inner class only one instance is stored in the database
+	// The fix (>= instead of > : currentCommitThreshold) leads to stable commit duplicates in the DB in case of multiple refactorings in the current commit
+	public boolean calculateStability(List<Integer> commitThresholds){
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/ProcessMetricTracker.java]: The method is called `calculateStability` but returns a boolean. Maybe a small doc here indicating what true and false means?

File: data-collection/src/main/java/refactoringml/db/Project.java [hunk 2]
```diff
@@ -26,8 +24,14 @@
 
 	private int commits;
 
-	//Collect instances of non-refactorings with different Ks e.g, 25,50,100 commits on a file without refactorings
+	//Collect instances of non-refactorings with different Ks e.g, 25,50,100 commits on a file without refactorings, this is for hibernate
 	private String commitCountThresholds;
+	//Collect instances of non-refactorings with different Ks e.g, 25,50,100 commits on a file without refactorings
+	@Transient
+	private List<Integer> commitCountThresholdsInt;
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/db/Project.java]: Why is this here? The list of Ks doesn't seem to be a characteristic of `Project` itself. I understand you tried to encapsulate this list of ints somewhere, otherwise you'd have to pass it all over, but maybe that's more semantic/clear than hiding it here.

File: data-collection/src/main/java/refactoringml/ProcessMetricTracker.java (hunk 1 above)
Comment (mauricioaniche) [reply data-collection/src/main/java/refactoringml/ProcessMetricTracker.java]: Maybe we should separate the method that returns whether this is stable from the one that update the currentCommitThreshold.

[Mixing "command and query" in a single method is usually not a good idea](https://en.wikipedia.org/wiki/Command–query_separation).

File: data-collection/src/main/java/refactoringml/ProcessMetricTracker.java [hunk 3]
```diff
@@ -74,6 +77,29 @@ public void resetCounter(CommitMetaData commitMetaData) {
...
+		for (Integer threshold : commitThresholds){
+			//1. Test if the class file was not refactored for the last K commits
+			if(isStable(threshold) &&
+					//2. Avoid duplicates:
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/ProcessMetricTracker.java]: maybe also extract these two conditions (i.e., `isStable(threshold)` and `(threshold > currentCommitThreshold || threshold == getCommitCounter())`) in two variables, each one with its own comments, so that we don't have to break the `if` in this strange manner just to write the comments.

File: data-collection/src/test/java/refactoringml/ProcessMetricTrackerTest.java [hunk 4]
```diff
@@ -151,9 +152,31 @@ public void countBugFixes() {
...
 			}
 
-			pm.reportCommit("bla bla " + (keywordHere) + "ble ble","Rafael", 10, 20);
+			pm.reportCommit("#" + (i + 2), "bla bla " + (keywordHere) + "ble ble","Rafael", 10, 20);
 		}
 
 		Assert.assertEquals(qty, pm.getCurrentProcessMetrics().bugFixCount);
 	}
+
+	@Test
```
Comment (mauricioaniche) [inline data-collection/src/test/java/refactoringml/ProcessMetricTrackerTest.java]: It is hard for me to review if the algorithm actually works, so I'm trusting on the (integration) tests. But maybe unit tests focused on the `calculateStability()` and `isStable()` will help us in getting more confidence.
